### ISSUE TITLE
Scrutinee dropped after if-let body

### ISSUE THREAD

**Issue by Aquam4n** [rust-lang/rust#133677]
Objects inside block expressions in the scrutinee of an if-let statement are not dropped before the body.

This is inconsistent both with declaring a temporary variable, e.g. `x` in test #2, and let-else behavior in test #3. 

I tried this code:

```rust
fn main() {
	struct S(i32);
	impl Drop for S {
		fn drop(&mut self) {
			println!("Dropped S {}", self.0)
		}
	}

	println!("--- test 1");
	{
		println!("Before if-let");
		if let _ = { S(1).0 } {
			println!("Inside body");
		}
	}

	println!("--- test 2");
	{
		println!("Before if-let");
		let x = { S(2).0 };
		if let _ = x {
			println!("Inside body");
		}
	}

	println!("--- test 3");
	{
		println!("Before let-else");
		let _ = ({ S(3).0 }) else { unreachable!() };
		println!("After let-else");
	}
}
```

I expected to see this happen:
```
--- test 1
Before if-let
Dropped S 1
Inside body
--- test 2
Before if-let
Dropped S 2
Inside body
--- test 3
Before let-else
Dropped S 3
After let-else
```

Instead, this happened:
```
--- test 1
Before if-let
Inside body
Dropped S 1
--- test 2
Before if-let
Dropped S 2
Inside body
--- test 3
Before let-else
Dropped S 3
After let-else
```

### Meta

Same behavior on both stable and nightly.

`rustc --version --verbose`:
```
rustc 1.82.0 (f6e511eec 2024-10-15)
binary: rustc
commit-hash: f6e511eec7342f59a25f7c0534f1dbea00d01b14
commit-date: 2024-10-15
host: x86_64-unknown-linux-gnu
release: 1.82.0
LLVM version: 19.1.1
```

**Comment by theemathas** [comment]
This is working as intended. For related discussion, see https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-if-let-scope.html and https://github.com/rust-lang/rust/issues/131154

**Comment by Aquam4n** [comment]
> This is working as intended. For related discussion, see https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-if-let-scope.html and [#131154](https://github.com/rust-lang/rust/issues/131154)

No. Allow me to elucidate. Temporaries in the scrutinee are not the same as temporaries inside a **block expression** in the scrutinee.

Take a look at:

```
fn main() {
	struct S(i32);
	impl Drop for S {
		fn drop(&mut self) {
			println!("Dropped S {}", self.0)
		}
	}

	println!("--- test 4");
	{
		println!("Before if-let");
		if let _ = { S(4).0 } {
			println!("Inside body");
		}
	}
	println!("--- test 5");
	{
		println!("Before if-let");
		if let _ = { let x = S(5).0; x } {
			println!("Inside body");
		}
	}
}
```

This prints

```
--- test 4
Before if-let
Inside body
Dropped S 4
--- test 5
Before if-let
Dropped S 5
Inside body
```

, it should print

```
--- test 4
Before if-let
Dropped S 4
Inside body
--- test 5
Before if-let
Dropped S 5
Inside body
```

**Comment by bjorn3** [comment]
Everything works they way you expect in the 2024 edition: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=c7245eb79cc6379d3cc19875401925ad So this is indeed fixed by https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-if-let-scope.html

**Comment by Aquam4n** [comment]
> Everything works they way you expect in the 2024 edition: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=c7245eb79cc6379d3cc19875401925ad So this is indeed fixed by https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-if-let-scope.html

Indeed, but it's not supposed to according to this description:

```
The 2024 Edition shortens the lifetime of the temporaries to the point where the then-block is completely evaluated or the program control enters the else block.
```

Maybe block expressions in the scrutinee were fixed by accident in the 2024 edition?

**Comment by bjorn3** [comment]
I think both https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-if-let-scope.html and https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-tail-expr-scope.html individually fixed this. Before the latter change all temporaries of the tail expression of a block had their lifetime extended to match the context within which the block appeared, so before https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-tail-expr-scope.html `let a = { S(1).0 }` is supposed to have the exact same lifetimes as `let a = S(1).0` and by extension `if let _ = { S(1).0 }` is supposed to have the same lifetimes as `if let _ = S(1).0`, which before https://doc.rust-lang.org/nightly/edition-guide/rust-2024/temporary-if-let-scope.html would drop `S` after the end of the if statement.

**Comment by est31** [comment]
The question seems to be answered. Can this be closed?

**Comment by est31** [comment]
Closing this as I'd say it's been answered.
